[#10] Validation 추가 및 Authorization

jjeda opened the pull request:
- AccountDto 객체에 Validation 추가
- Validation 검증을 위한 테스트코드 추가
- 관리자와 유저 account 컨트롤러 분리
- Role에 따른 Authrization 범위 설정
- Authrization 테스트 코드 추가
- TODO : 유저가 수정, 삭제, 조회 할때 현재 사용자 정보를 불러와 해당 사용자가 아닐 경우 401

jjeda committed:
[#10] Validation 추가 및 Authorization

- AccountDto 객체에 Validation 추가
- Validation 검증을 위한 테스트코드 추가
- 관리자와 유저 account 컨트롤러 분리
- Role에 따른 Authrization 범위 설정
- Authrization 테스트 코드 추가

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/Service/AccountService.java:
```diff
@@ -54,16 +55,19 @@ public PagedResources findAllAccountWithValidation(Boolean isDeleted, Pageable p
     }
 
+    @Transactional
     public void deleteAccountByAdmin(Long id) {
```
delete에 대한 메소드가 위의 `deleteAccount`와 동작이 중복으로 보입니다~
보통은 관리자에 의한 탈퇴가 아니라 이런 경우는 어뷰징 등의 행위로 계정정지를 당하는 경우로 볼 수 있는데요.

1. 관리자에 의해서 밴을 당하는 거라면 `이 사람은 관리자에 대해 정지당한 사람이다` 라는 플래그를 따로 정의해주는게 좋아보입니다.
2. 굳이 By~~를 메소드에 붙여줄 필요는 없습니다. 누가 행위를 하는지는 Security에서 권한명시로 충분할 것 같습니다.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/Service/AccountService.java:
```diff
@@ -54,16 +55,19 @@ public PagedResources findAllAccountWithValidation(Boolean isDeleted, Pageable p
         account.setDeleteFlag();
     }
 
+    @Transactional
```
[팁] `@Transactional`을 사용할 때에도 이 것이 꼭 필요한지 고민해보는 것이 좋습니다.
트랜잭션을 거는 행위도 부하가 드는 작업이기에 만약 동일한 유저에 대해 동시적으로 수정요청이 들어오지 않는다는게 보장된다면 굳이 걸어줄 필요가 없습니다.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -26,55 +22,39 @@ public AccountController(AccountService accountService) {
-    @GetMapping("/admin/{id}")
-    public ResponseEntity getAccountByAdmin(@PathVariable Long id) {
-        Optional<Account> account = accountService.getAccountByAdmin(id);
+        if (account.isEmpty()) return ResponseEntity.notFound().build();
```
[팁] 가끔은 한 줄로 작성하는게 (타인이 봤을 때에도) 가독성이 좋을지, 괄호를 넣어주는게 가독성이 좋을 지 고민해보시는 것도 좋습니다.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -26,55 +22,39 @@ public AccountController(AccountService accountService) {
     }
 
     @DeleteMapping("/{id}")
     public ResponseEntity deleteAccount(@PathVariable Long id) {
```
이 클래스의 이름은 AccountController인데 그렇다면 이 행위를 하는 주체는 유저 본인일것이라고 예상합니다.
userId는 세션에서 가져올 수 있을 텐데 파라미터로 받지 않아도 될 것 같습니다.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AdminAccountController.java:
```diff
@@ -0,0 +1,56 @@
+     * @param isDeleted : Account 가 삭제됐을 때 true 값을 갖는 flag
+     */
+    @GetMapping
+    public ResponseEntity getAllAccountWithValidation(@RequestParam Boolean isDeleted, Pageable pageable, PagedResourcesAssembler<Account> pagedResourcesAssembler) {
```
밴시킨 유저 목록을 조회하는 것은 의미가 있어보이는데요~ 일반 유저까지 조회할 필요가 있을까요?
그리고 단순히 목록 조회보다는 검색조건을 넣어주는게 좋아보입니다.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AdminAccountController.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity getAccountByAdmin(@PathVariable Long id) {
```
'Admin'AccountController이므로 ByAdmin은 빼도 의미를 알 수 있을 것 같습니다

f-lab-dev reviewed (commented):
확인완료했습니다~ 팁을 드리자면 어플이나 웹의 화면을 대략적으로 스케치해서 동작을 모두 정의해보시는 것도 추천드립니다. 그러면 어떤 기능이 필요한지 정의가 나올테니까요~

jjeda committed:
[#10] AccountStatus 추가 & 사용자 로직 수정

- isDeleted Flag 삭제 -> 상태를 다양화 할 수 있는 AccountStatus enum 추가
- 상태전이 코드 추가
- 사용자가 @Pathvariable을 통해 endpoint로 접근하는 것이 아닌 현재사용자 정보를 받아와 접근하게 수정
- 기타 : import에 check style 적용, Appconfig 이름 역할에 맞게 수정

jjeda commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -29,32 +38,36 @@ public ResponseEntity createAccount(@RequestBody @Valid AccountDto requestAccoun
-        if (account.isEmpty()) return ResponseEntity.notFound().build();
+    @GetMapping
+    public ResponseEntity getAccount(@AuthenticationPrincipal User currentUser) {
+        Optional<Account> account = accountService.getAccount(currentUser.getUsername());
```
@f-lab-dev 
이번 commit에 가장 큰 고민이었던 부분인데요..
`@AuthenticationPrincipal`애노태이션을 사용하면 (UserDetailsService를 구현한 AccountService의)loadByUsername이 리턴하는 객체를 받아 오는 것으로 알고있습니다.
저는 Spring security가 제공하는 User를 리턴하도록 코드를 작성하였는데
그렇기 때문에 (Account가 아니기 때문에) id값이 없어 Username으로 DB에서 받아왔는데요

이런식으로 username으로 사용자를 받아와도 되는지 아니면 어댑터 패턴을 사용하여 UserDetails이랑 Account를 연결해주는 어댑터를 만들어야 하는지 궁금합니다!

f-lab-dev replied:
userName이 유니크하다는 전제하에 그렇게 불러와도 됩니다. (username에는 인덱스를 걸어야겠죠)
무조건 스프링의 기술을 사용할 필요는 없습니다~ 커스텀하게 구현할만한 곳은 커스텀하게 구현하는게 좋을 것 같습니다.

f-lab-dev reviewed (commented):
이번 변경은 깔끔하네요~

jjeda committed:
[#10] Current User 로직 수정

- AccountAdapter 클래스 구현 ( extends User)
- 어댑터 패턴 적용

jjeda commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -39,35 +40,25 @@ public ResponseEntity createAccount(@RequestBody @Valid AccountDto requestAccoun
-            return ResponseEntity.notFound().build();
-        }
+    public ResponseEntity getAccount(@AuthenticationPrincipal AccountAdapter accountAdapter) {
+        Account account = accountAdapter.getAccount();
```
어제 고민한 이부분에 Adapter를 적용한 이유는
42~43 번라인처럼 현재유저를 받아와서 다시 Account 객체를 바꾸는 과정에 DB를 타기 때문에
DB를 타지 않는 방법으로 로직을 수정했습니다.
추후에 상품 주문시 해당 상품의 주인( Account)을 지정해줘야 하는 로직이 들어 갈때도 DB를 통해 Account를 받아온는 것은 비효율적이라고 생각했습니다

jjeda replied:
@f-lab-dev 
해당 이슈 관련 처음으로.. ㅎ 포스팅을 해봤는데
혹시 이것도 리뷰 부탁드려도 될까요~
https://jjeda.tistory.com/7 여깁니다!

f-lab-dev replied:
@jjeda 만약 개인 정리용이 아닌 지식 공유용이라면 `draw.io`등을 이용해서 구조를 대략적으로 그림을 그려주는것도 좋은 방법입니다. 개인적인 고민을 나타내는 글에는 기울임처리를 한 것도 좋았지만 거기에 bold처리까지 해주면 가독성이 좋아질 것 같습니다. (고민한 내용은 기존의 글보다 더 눈에 잘 들어오게 해주면 좋을 것 같아요)

어노테이션 같은 소스코드를 나타낼때는 "`@AuthenticationPrincipal`을 사용해서" 처럼 소스코드에 대한 별도의 하이라이팅을 주면 좋을 것 같고요~

제목도 내용을 보니 깔끔하게 사용자를 불러올 수 있는 팁을 담고있데 제목과 글의 초반에 이에 대한 내용임을 암시할 수 있도록 해주는 것도 좋을 것 같습니다~

글을 읽는 사람이 다음 단계에 대해서도 생각할 수 있도록 more을 달아준 것이 매우 좋아보이네요

f-lab-dev replied:
그리고 Adapter를 적용한 이유가 합리적이네요. "DB에 대한 부하를 줄여준다."
적용하시길 잘 한 것 같습니다.

jjeda commented on src/main/java/me/jjeda/mall/accounts/dto/AccountAdapter.java:
```diff
@@ -0,0 +1,31 @@
+package me.jjeda.mall.accounts.dto;
```
dto 패키지에 AccountAdapter를 위치시켰는데 적절한지 궁금합니다..

f-lab-dev replied:
AccountAdapter는 dto가 아니기에 적절하지는 않은 것 같네요
domain 쪽이 더 적절해보입니다

jjeda committed:
[#10] 현재 사용자 로직수정

- expression = account SPEL을 통해 간결하게 변경
- 간결함, 가독성을 위해 CurrentUser 애노태이션 정의

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -3,6 +3,7 @@
 import me.jjeda.mall.accounts.Service.AccountService;
 import me.jjeda.mall.accounts.domain.Account;
 import me.jjeda.mall.accounts.domain.AccountStatus;
```
import 구문을 보면 unused import 구문이 많은데 이들은 삭제해주는 것이 좋습니다~

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/Service/AccountService.java:
```diff
@@ -74,12 +72,7 @@ public UserDetails loadUserByUsername(String username) throws UsernameNotFoundEx
                 .orElseThrow(() -> new UsernameNotFoundException(username));
 
-        return new User(account.getEmail(), account.getPassword(), authorities(account.getAccountRole()));
+        return new AccountAdapter(account);
```
`AccountAdapter`는 `Account`와 관계가 깊으므로 new 해서 객체를 생성하기보다는 static으로 from같은 팩토리메소드를 정의해주는 것이 좋을 것 같습니다.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -40,26 +41,21 @@ public ResponseEntity createAccount(@RequestBody @Valid AccountDto requestAccoun
     @DeleteMapping
-    public ResponseEntity deleteAccount(@AuthenticationPrincipal AccountAdapter accountAdapter) {
-        Account account = accountAdapter.getAccount();
+    public ResponseEntity deleteAccount(@CurrentUser Account account) {
```
이 API는 사용자 관점에서의 API이기 때문에 이름도 사용자 관점에서 지어주는게 좋을 것 같습니다.
탈퇴와 관련된 이름으로 지어주는게 좋을 것 같네요.

f-lab-dev commented on src/main/java/me/jjeda/mall/accounts/controller/AccountController.java:
```diff
@@ -40,26 +41,21 @@ public ResponseEntity createAccount(@RequestBody @Valid AccountDto requestAccoun
-
-        account = accountService.updateAccount(account.getId(), accountDto);
+    public ResponseEntity updateAccount(@RequestBody @Valid AccountDto accountDto, @CurrentUser Account account) {
+        Account newAccount = accountService.updateAccount(account.getId(), accountDto);
```
`newAccount`라는 단어는 새로 생성된 유저같은 느낌이 드네요. 유지보수성 측면에서 updatedAccount라고 지어주는게 좋을 것 같아요~

f-lab-dev commented on src/main/java/me/jjeda/mall/common/CurrentUser.java:
```diff
@@ -0,0 +1,14 @@
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+@AuthenticationPrincipal(expression = "account")
+public @interface CurrentUser {
```
[팁] 이거는 고민을 해볼만한 문제인데요.
단순히 오버라이드해서 커스텀화한 어노테이션이 많을수록 유지보수하는 관점에서 혼란을 줄 가능성이 생깁니다. 
이 정도는 그냥 어노테이션 소스코드 보고 "아 이거는 어떤거구나~" 하고 볼 수 있지만 개수가 너무 늘어나면 혼란이 올 수도 있습니다.

f-lab-dev reviewed (commented):
리뷰완료했습니다~

f-lab-dev commented:
이제 리뷰가 엄청 많이 남게되었는데요~ 이 이슈의 하위 이슈를 좀 더 세분화해서 등록하고 이 브랜치로 PR을 날리면 좋을 것 같네요.

jjeda committed:
[#16] Account 관련 수정 및 개선

- unused import 삭제
- 변수명 및 매서드명 수정

jjeda committed:
Merge pull request #18 from f-lab-edu/feature/16

[#16] Account 관련 수정 및 개선
